From: David Steele <dsteele@gmail.com>
Date: Thu, 16 Apr 2015 21:01:01 -0400
Subject: cloudprint.py: retry failed print jobs

the first attempt to call CUPS printFile() is failing. I don't know
why, but the job completes with a retry. Add a rudimentary retry
mechanism that will not delete the job from Cloud Print for a first
failure.

Conflicts:
	cloudprint/cloudprint.py
---
 cloudprint/cloudprint.py | 17 ++++++++++++++---
 1 file changed, 14 insertions(+), 3 deletions(-)

diff --git a/cloudprint/cloudprint.py b/cloudprint/cloudprint.py
index 7812539..a63eff6 100755
--- a/cloudprint/cloudprint.py
+++ b/cloudprint/cloudprint.py
@@ -56,6 +56,10 @@ FAIL_RETRY = 60
 # how often, in seconds, to send a keepalive character over xmpp
 KEEPALIVE = 600.0
 
+# failed job retries
+RETRIES = 1
+num_retries = 0
+
 LOGGER = logging.getLogger('cloudprint')
 LOGGER.setLevel(logging.INFO)
 
@@ -371,8 +375,8 @@ def sync_printers(cups_connection, cpp):
     for printer_name in remote_printer_names - local_printer_names:
         remote_printers[printer_name].delete()
 
-
 def process_job(cups_connection, cpp, printer, job):
+    global num_retries
     try:
         pdf = cpp.auth.session.get(job['fileUrl'], stream=True)
         tmp = tempfile.NamedTemporaryFile(delete=False)
@@ -391,9 +395,16 @@ def process_job(cups_connection, cpp, printer, job):
 
         cpp.finish_job(job['id'])
 
+        num_retries = 0
+
     except Exception:
-        cpp.fail_job(job['id'])
-        LOGGER.exception('ERROR ' + job['title'].encode('unicode-escape'))
+        if num_retries >= RETRIES:
+            num_retries = 0
+            cpp.fail_job(job['id'])
+            LOGGER.error('ERROR ' + job['title'].encode('unicode-escape'))
+        else:
+            num_retries += 1
+            LOGGER.info('Job %s failed - Will retry' % job['title'].encode('unicode-escape'))
 
 
 def process_jobs(cups_connection, cpp):
